draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,69 +60,304 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+chosen `vdaf_type`. The aggregators MUST pass `vdaf_data` to VDAF initialiser,
+based on the chosen `vdaf_type`.
+
+The definition of Time, Duration, Url, QueryType follow those in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.
+
+Note that the parameters that are not necessarily tied to a task (for e.g.
+collector hpke config), and secrets that should not be known by clients, like
+`vdaf_verify_key`, MUST still be exchanged out-of-band among aggregators.
 
 # Client Behavior
 
 > TODO(wangshan) Say how the Client constructs the extension body. Does the
 > extension change anything else about the upload flow?
 
+The client should know whether `task-prov` extension will be used and all
+parameters required for `TaskConfig` prior to constructing the extension
+body, either out-of-band from aggregators, or from information already saved on
+client.
+
+To use this extension, the client should first decide what is the task ID, see
+{{construct-task-id}}.
+
+## Construct task ID {#construct-task-id}
+
+For `task-prov` extension, a DAP task is not created before distributing task
+configuration to clients. Therefore, a DAP task ID may not be available to
+clients, aggregators and collector before uploading. Aggregator or collector
+MAY still choose a task ID for this task and deliver it to all clients along
+with `TaskConfig`. In this case, client MUST use that task ID. This arrangement
+SHOULD be agreed by aggregators, collector and clients out-of-band.
+
+Alternatively, task ID can be constructed from `task_info` and any other part
+of task configuration, as long as the generated task ID is determinstic and
+stay consistent across all parties. For example, the task ID can be a SHA256
+hash of the entire serialized `TaskConfig`. In this case, the `task_info`
+serves as an unique byte array to distinguish different tasks that share the
+same task parameters. Task ID can be constructed on server side and then
+distributed to clients, or constructed independently on clients and servers.
+When constructed independently, the mechanism used for creating the task ID
+must be known to both clients and the collector.
+
+## Construct extension body
+
+Client constructs this extension during the upload flow, after hpke config
+(see update flow and hpke-config in {{?DAP=I-D.draft-ietf-ppm-dap-01}}.). Note
+that if task ID is not available at time of hpke config query, the client
+should use `[aggregator]/hpke_config` API without specifying a `task_id`.
+Client typically sets `extension_type` to `task-prov` codepoint in
+`ReportMetadata`'s extenion field, and save the encoded `TaskConfig` in
+`extension_data` field.
 
 # Leader Behavior
 
 > TODO(wangshan) Say what the Leader does when it receives a report with this
 > extension. What do Leaders do if they *don't* support this extension? Is it
 > safe to ignore it? How are the upload, aggregate, and collect flows impacted?
 
+Leader should have saved any information that is not always tied to a
```

Comment:
I wonder if we should briefly mention all the "secret" parameters in a task that cannot be agreed through the extension-based task provisioning. On top of my head, these include `vdaf_verify_key`, `collector_hpke_config`, and `aggregator_auth_tokens`

Reply:
that is mentioned just above the client behaviour section